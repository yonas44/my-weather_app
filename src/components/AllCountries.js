import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { MdOutlineExplore, MdOutlineSmsFailed } from 'react-icons/md';
import Country from './country';
import load from '../assets/load.gif';
import paginate from './utils/paginate';

const AllCountries = () => {
  const [currentPage, setPage] = useState(1);
  const countriesStore = useSelector((state) => state.countries);
  const {
    loading, countries, sucess, selected: { continent }, errorInfo, countriesPerPage,
  } = countriesStore;
  const filteredCountries = countries.filter((country) => country.region === continent);
  const paginatedCountries = paginate(filteredCountries, currentPage, countriesPerPage);

  const handlePageChange = (input) => {
    if (input === 'prev') return setPage((prevState) => prevState - 1);
    return setPage((prevState) => prevState + 1);
  };

  const navigate = useNavigate();
  useEffect(() => {
    if (continent === 'All') navigate('/');
  }, [continent, navigate]);

  return (
    <main className="countries-wrapper">
      <div className={`${continent} bigger-continent-img`}>
        <h2 id="continent-title">
          {continent}
        </h2>
        <p style={{ fontStyle: 'italic' }}>
          {`Home of ${filteredCountries.length} countries!`}
        </p>
        <a
          id="icon-wrapper"
          target="_blank"
          rel="noopener noreferrer"
          href={`https://en.wikipedia.org/wiki/${continent}`}
        >
          <MdOutlineExplore id="explore-icon" />
        </a>
      </div>
      {loading && (
        <div className="load-wrapper-all-countries">
          <img
            style={{ width: '150px', height: '150px' }}
            src={load}
            alt="loading"
          />
        </div>
      )}
      { sucess ? (
        <section className="countries-section">
          <div className="countries-holder">
            {paginatedCountries.map((country) => (
              <Country
                key={country.name.common}
                name={country.name.common}
                flag={country.flags.svg}
                timezone={country.timezones[0]}
                area={country.area}
              />
            ))}
          </div>
          <div className="pagination-holder">
            <span id="page-number">{`${currentPage}/ ${Math.ceil(filteredCountries.length / countriesPerPage)}`}</span>
            <div className="buttons-holder">
              <button
                type="button"
                disabled={currentPage <= 1}
                onClick={() => handlePageChange('prev')}
                className="pagination-btns"
              >
                Previous
              </button>
              <button
                type="button"
                disabled={currentPage >= Math.ceil(filteredCountries.length / countriesPerPage)}
                onClick={() => handlePageChange('next')}
                className="pagination-btns"
              >
                Next
              </button>
            </div>
          </div>
        </section>
      ) : (
        <div id="allCountries-error">
          <MdOutlineSmsFailed style={{ color: 'red', scale: '2' }} />
          <span>{errorInfo}</span>
        </div>
      )}
    </main>
  );
};

export default AllCountries;
