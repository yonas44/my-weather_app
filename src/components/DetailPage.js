import React, { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { MdOutlineSmsFailed } from 'react-icons/md';
import load from '../assets/load.gif';

const DetailPage = () => {
  const {
    loading, weather, sucess, errorInfo,
  } = useSelector((state) => state.weather);
  const countriesInfo = useSelector((state) => state.countries);
  const selected = countriesInfo.selected.country;
  const country = countriesInfo.countries.filter(
    (card) => card.name.common === selected,
  )[0] ?? {};

  const naviagate = useNavigate();
  useEffect(() => {
    if (countriesInfo.countries.length === 0) naviagate('/');
  }, [countriesInfo.countries.length, naviagate]);

  return (
    <div className="detail-page">
      <div className="country-header">
        {loading ? (
          <div className="load-wrapper">
            <img
              style={{ width: '100px', height: '100px' }}
              src={load}
              alt="loading"
            />
          </div>
        ) : (
          ''
        )}
        <img id="country-header-img" src={country.flags?.png} alt="flag" />
        <>
          {sucess ? (
            <div className="header-info">
              <p id="country-name">{country.name?.common}</p>
              <div className="weather-info">
                <div id="top-section">
                  <img src={weather?.condition.icon} alt="weather-icon" />
                  <p>{weather.condition.text}</p>
                </div>
                <div id="bottom-section">
                  <p>{`${weather.maxtemp}째C / ${weather.mintemp}째C`}</p>
                  <p style={{ fontSize: '0.85rem' }}>{`Humidity: ${weather.hum}%`}</p>
                </div>
              </div>
            </div>
          ) : (
            <div id="detail-page-error">
              <MdOutlineSmsFailed style={{ color: 'white', scale: '2' }} />
              <span id="detail-error-message">{errorInfo}</span>
            </div>
          )}
        </>
      </div>
      <div className="country-detail">
        <p>
          <strong>
            Capital:
          </strong>
          {` ${country.capital ? country.capital[0] : ''}`}
        </p>
        <p>
          <strong>
            Population:
          </strong>
          {` ${country.population?.toLocaleString()}`}
        </p>
        <p>
          <strong>
            Location:
          </strong>
          {` ${country.latlng ? country.latlng[0] : ''}째N and ${country.latlng ? country.latlng[0] : ''}째E`}
        </p>
        <p>
          <strong>
            Subregion:
          </strong>
          {` ${country.subregion ?? ''}`}
        </p>
      </div>
    </div>
  );
};

export default DetailPage;
