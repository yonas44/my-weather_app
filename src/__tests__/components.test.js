import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import AllCountries from '../components/AllCountries';
import Header from '../components/header';
import store from '../feature/configureStore';
import DetailPage from '../components/DetailPage';
import Country from '../components/country';

describe('AllCountries-component renders', () => {
  test('It renders in the screen', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <AllCountries />
        </BrowserRouter>
      </Provider>,
    );

    expect(screen.getByAltText(/loading/)).toBeInTheDocument();
  });

  test('It matchs the snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <AllCountries />
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('Header-component renders', () => {
  test('It renders in the screen', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </Provider>,
    );

    expect(screen.getByText(/All countries/)).toBeInTheDocument();
  });

  test('It matchs the snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <Header />
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('DetailPage-component renders', () => {
  test('It renders in the screen', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <DetailPage />
        </BrowserRouter>
      </Provider>,
    );

    expect(screen.getByText(/Capital/)).toBeInTheDocument();
  });

  test('It matchs the snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <DetailPage />
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('SingleCountry-component renders', () => {
  test('It renders in the screen', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Country />
        </BrowserRouter>
      </Provider>,
    );

    expect(screen.getByText(/Size/)).toBeInTheDocument();
  });

  test('It matchs the snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <Country />
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
