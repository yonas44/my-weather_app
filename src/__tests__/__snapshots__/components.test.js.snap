// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllCountries-component renders It matchs the snapshot 1`] = `
<div
  className="countries-wrapper"
>
  <div
    className="All bigger-continent-img"
  >
    <h2
      id="continent-title"
    >
      All
    </h2>
    <p
      style={
        Object {
          "fontStyle": "italic",
        }
      }
    >
      Home of 0 countries!
    </p>
    <a
      href="https://en.wikipedia.org/wiki/All"
      id="icon-wrapper"
      rel="noopener noreferrer"
      target="_blank"
    >
      <svg
        fill="currentColor"
        height="1em"
        id="explore-icon"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm-5.5-2.5l7.51-3.49L17.5 6.5 9.99 9.99 6.5 17.5zm5.5-6.6c.61 0 1.1.49 1.1 1.1s-.49 1.1-1.1 1.1-1.1-.49-1.1-1.1.49-1.1 1.1-1.1z"
        />
      </svg>
    </a>
  </div>
  <div
    className="countries-holder"
  />
</div>
`;

exports[`DetailPage-component renders It matchs the snapshot 1`] = `
<div
  className="detail-page"
>
  <div
    className="country-header"
  >
    <img
      alt="flag"
      id="country-header-img"
    />
    <div
      className="header-info"
    >
      <p
        id="country-name"
      />
      <div
        className="weather-info"
      >
        <div
          id="top-section"
        >
          <img
            alt="weather-icon"
          />
          <p />
        </div>
        <div
          id="bottom-section"
        >
          <p>
            째C / 째C
          </p>
          <p
            style={
              Object {
                "fontSize": "0.85rem",
              }
            }
          >
            Humidity: %
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="country-detail"
  >
    <p>
      <strong>
        Capital:
      </strong>
       
    </p>
    <p>
      <strong>
        Population:
      </strong>
       undefined
    </p>
    <p>
      <strong>
        Location:
      </strong>
       째N and 째E
    </p>
    <p>
      <strong>
        Subregion:
      </strong>
       
    </p>
  </div>
</div>
`;

exports[`Header-component renders It matchs the snapshot 1`] = `
<header>
  <div
    style={
      Object {
        "color": "white",
        "textDecoration": "none",
      }
    }
  />
  <span>
    Continents
  </span>
  <div
    className="search-wrapper"
  >
    <div
      className="search-field"
    >
      <input
        id="country-input"
        onChange={[Function]}
        onKeyUp={[Function]}
        placeholder="Enter a country name.."
        required={true}
        type="text"
        value=""
      />
    </div>
    <select
      id="filter-select"
    >
      <option
        onClick={[Function]}
        style={
          Object {
            "fontWeight": "bolder",
          }
        }
      >
        All
      </option>
      <option
        onClick={[Function]}
      >
        Africa
      </option>
      <option
        onClick={[Function]}
      >
        Americas
      </option>
      <option
        onClick={[Function]}
      >
        Asia
      </option>
      <option
        onClick={[Function]}
      >
        Europe
      </option>
    </select>
  </div>
</header>
`;

exports[`SingleCountry-component renders It matchs the snapshot 1`] = `
<button
  className="country-card"
  onClick={[Function]}
  type="button"
>
  <a
    href="/country"
    onClick={[Function]}
  >
    <img
      alt="flag"
      id="flag"
      src={null}
    />
    <div
      className="country-info"
    >
      <p
        id="timezone"
      />
      <p
        id="country-name"
      />
      <p
        id="country-size"
      >
        Size: undefinedKm2
      </p>
    </div>
  </a>
</button>
`;
